e2e/tests/wasm/grandpa_test.go

```diff
@@ -255,6 +204,157 @@ func (s *GrandpaTestSuite) TestMsgTransfer_Succeeds_GrandpaContract() {
+		Msg:      []byte("{}"),
+	}
+
+	s.ExecuteAndPassGovV1Proposal(ctx, &message, cosmosChain, cosmosWallet)
+
+	clientState, err := s.QueryClientState(ctx, cosmosChain, "08-wasm-0")
+	s.Require().NoError(err)
+
+	wasmClientState, ok := clientState.(*wasmtypes.ClientState)
+	s.Require().True(ok)
+
+	s.Require().Equal(newCodeHashBz, wasmClientState.CodeHash)
+}
+
+func (s *GrandpaTestSuite) TestMsgMigrateContract_ContractError_GrandpaContract() {
+	ctx := context.Background()
+
+	chainA, chainB := s.GetGrandpaTestChains()
+
+	polkadotChain := chainA.(*polkadot.PolkadotChain)
+	cosmosChain := chainB.(*cosmos.CosmosChain)
+
+	// we explicitly skip path creation as the contract needs to be uploaded before we can create clients.
+	r := s.ConfigureRelayer(ctx, polkadotChain, cosmosChain, nil, func(options *interchaintest.InterchainBuildOptions) {
+		options.SkipPathCreation = true
+	})
+
+	s.InitGRPCClients(cosmosChain)
+
+	var err error
+	cosmosWallet := s.CreateUserOnChainB(ctx, testvalues.StartingTokenAmount)
+
+	file, err := os.Open("../data/ics10_grandpa_cw.wasm")
+	s.Require().NoError(err)
+
+	codeHash := s.PushNewWasmClientProposal(ctx, cosmosChain, cosmosWallet, file)
+
+	s.Require().NotEmpty(codeHash, "codehash was empty but should not have been")
+
+	eRep := s.GetRelayerExecReporter()
+
+	// Set client contract hash in cosmos chain config
+	err = r.SetClientContractHash(ctx, eRep, cosmosChain.Config(), codeHash)
+	s.Require().NoError(err)
+
+	// Ensure parachain has started (starts 1 session/epoch after relay chain)
+	err = testutil.WaitForBlocks(ctx, 1, polkadotChain)
+	s.Require().NoError(err, "polkadot chain failed to make blocks")
+
+	pathName := s.GetPathName(0)
+
+	err = r.GeneratePath(ctx, eRep, cosmosChain.Config().ChainID, polkadotChain.Config().ChainID, pathName)
+	s.Require().NoError(err)
+
+	// Create new clients
+	err = r.CreateClients(ctx, eRep, pathName, ibc.DefaultClientOpts())
+	s.Require().NoError(err)
+	err = testutil.WaitForBlocks(ctx, 1, cosmosChain, polkadotChain) // these 1 block waits seem to be needed to reduce flakiness
+	s.Require().NoError(err)
+
+	// Do not start the relayer
+
+	// This contract is a dummy contract that will always fail migration.
+	// Other entry points are unimplemented.
+	migrateFile, err := os.Open("../data/migrate_error.wasm.gz")
+	s.Require().NoError(err)
+
+	// First Store the code
+	newCodeHashHex := s.PushNewWasmClientProposal(ctx, cosmosChain, cosmosWallet, migrateFile)
+	s.Require().NotEmpty(newCodeHashHex, "codehash was empty but should not have been")
+
+	newCodeHashBz, err := hex.DecodeString(newCodeHashHex)
+	s.Require().NoError(err)
+
+	// Attempt to migrate the contract
+	message := wasmtypes.MsgMigrateContract{
+		Signer:   authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+		ClientId: "08-wasm-0",
+		CodeHash: newCodeHashBz,
+		Msg:      []byte("{}"),
+	}
+
+	err = s.ExecuteGovV1Proposal(ctx, &message, cosmosChain, cosmosWallet)
+	s.Require().Error(err)
```

Comment:
Maybe we should make a granular assertion about the contents or type of the error here? E.g. this could be a timeout to pass the proposal, or a codec panicking in the chain binary.

Reply:
done